File: two_pointers/trap_rainwater.py
from typing import List


class Solution:
    
    # using 1 loop and no memory
    def trap(self, height: List[int]) -> int:
        lmax, rmax = height[0], height[-1]
        output = 0
        l, r = 0, len(height) - 1
        while l < r:
            if height[l] < height[r]:
                lmax = max(lmax, height[l])
                trap = lmax - height[l]
                output += trap if trap > 0 else 0 
                l += 1
            else:
                rmax = max(rmax, height[r])
                trap = rmax - height[r]
                output += trap if trap > 0 else 0 
                r -= 1
        return output
    

    # using 2 loops and memory
    def trap2(self, height: List[int]) -> int:
        left = [0] * len(height)
        right = [0] * len(height)
        lmax, rmax = height[0], height[-1]
        for i in range(len(height)):
            lmax = max(lmax, height[i])
            left[i] = lmax
            revidx = len(height) - i - 1
            rmax = max(rmax, height[revidx])
            right[revidx] = rmax
        
        output = 0
        for i in range(len(height)):
            trap = min(left[i], right[i]) - height[i]
            output += trap if trap > 0 else 0
        return output

File: old/two_pointers/trap_rain_water.js
// Solution 1 -> O(n) & O(n)
/*
var trap = function(height) {
  const maxLeftList = new Array(height.length).fill(0);
  const maxRightList = new Array(height.length).fill(0);
  let res = 0;
  let maxLeft = 0;
  let maxRight = 0;
  for(let i=1; i<height.length; i++) {
      maxLeft = Math.max(maxLeft, height[i-1]);
      maxLeftList[i] =  maxLeft;
  }
  for(let i=height.length-2; i>=0; i--) {
      maxRight = Math.max(maxRight, height[i+1]);
      maxRightList[i] =  maxRight;
  }
  for(const i in height) {
      const calculatedRes = (Math.min(maxLeftList[i], maxRightList[i])) - height[i];
      res += calculatedRes > 0 ? calculatedRes : 0;
  }
  return res;
};
*/

// Solution 2
var trap = function(height) {
  if(height.length == 0) return 0;
  let res = 0;
  let left = 0;
  let right = height.length - 1;
  let maxLeft = height[left];
  let maxRight = height[right];
  while(left < right) {
      if(height[left] < height[right]) {
          left++;
          maxLeft = Math.max(maxLeft, height[left]);
          res += maxLeft - height[left];
      } else {
          right--;
          maxRight = Math.max(maxRight, height[right]);
          res += maxRight - height[right];
      }
  }
  return res;
};

console.log(trap([0,1,0,2,1,0,1,3,2,1,2,1]));